### PR TITLE
Let Run return an error to be emitted

### PR DESCRIPTION
**Pull request by keks**
This addresses #86 and @whyrusleeping's comments in #87.

### CODE REVIEW

File: command.go [hunk 1]
```diff
@@ -104,7 +104,16 @@ func (c *Command) Call(req *Request, re ResponseEmitter, env Environment) {
...
-	cmd.Run(req, re, env)
+	err = cmd.Run(req, re, env)
+	if err != nil {
+		if cmderr, ok := err.(cmdkit.Error); ok {
+			err = re.Emit(cmderr)
+			if err != nil {
+				log.Error("error emitting cmd error:", err)
+			}
+		}
+		re.SetError(err, cmdkit.ErrNormal)
```
Comment (Stebalien) [inline command.go]: Won't this emit the error twice?

Comment (keks) [reply command.go]: You're right, fixed now

File: command.go [hunk 2]
```diff
@@ -104,7 +104,18 @@ func (c *Command) Call(req *Request, re ResponseEmitter, env Environment) {
 		}
 	}
 
-	cmd.Run(req, re, env)
+	err = cmd.Run(req, re, env)
+	if err != nil {
+		if cmderr, ok := err.(cmdkit.Error); ok {
```
Comment (Stebalien) [inline command.go]: IIRC, we also need to check if this is a pointer to a `cmdkit.Error`. Right?

File: command.go [hunk 3]
```diff
@@ -104,7 +104,18 @@ func (c *Command) Call(req *Request, re ResponseEmitter, env Environment) {
...
+			err = re.Emit(cmderr)
+			if err != nil {
+				log.Error("error %q emitting cmd error %q on request %#v", err, cmderr, req)
```
Comment (Stebalien) [inline command.go]: I'm worried we'll hit this every time a client disconnects. What do you think?

File: command.go (hunk 2 above)
Comment (keks) [reply command.go]: *sigh* you're right. We should figure something out to make this less of a hassle - see https://github.com/ipfs/go-ipfs-cmdkit/issues/19

File: command.go (hunk 3 above)
Comment (keks) [reply command.go]: Hmm, a client disconnect shouldn't trigger an error in Run in the first place, so this entire branch shouldn't trigger - or am I missing something?

Comment (Stebalien) [reply command.go]: Client disconnects -> context gets canceled -> Run returns the context cancellation error.

Comment (keks) [reply command.go]: Hmm, but usually the server will write their response and return. What you write only happens if the server doesn't expect the client to exit, which is a rightful error IMHO.
If the Run function sends data and expects that the client might have closed the connection by now, they should not return context cancellation errors.

Comment (Stebalien) [reply command.go]: > If the Run function sends data and expects that the client might have closed the connection by now, they should not return context cancellation errors.

Telling whether or not the client has gone away can sometimes be a bit tricky and we definitely don't want to return "success" to the user if we *think* the client has disconnected but it hasn't.

Currently, we *always* return an error if we abort the request due to a client disconnect.

Comment (keks) [reply command.go]: Okay, so nothing changes then, right?

Comment (Stebalien) [reply command.go]: I guess in *this* situation, we're unlikely to return a cmdkit Error so we shouldn't hit this log message.

Although, I still don't think that this is a true error case (for the daemon) as the client could have disconnected after run completed but before we try to send the value back. Really, I think that's the only possible way to hit this log.

**Comment by Stebalien** [comment]
Merged in #112

**Pull request closed without merging**